### DISCUSSION TITLE
Versioning DB schema with Remult

### DISCUSSION THREAD

**Discussion by amrit92** [noam-honig/radweb#162] [General]
Would like to start by saying thanks for Remult, its a very simple concept and makes implementation a breeze. I'd like to know how a Database schema version is maintained (Postgres specifically) and if not what are the downsides. From what I've been playing around with, it seems like changes to model are immediately reflected on the database and there's no easy way of rolling back if needed except through code.
Also are there any examples of production usage of Remult? That'd solidify my confidence.

**Comment by noam-honig** [comment]
Hi @amrit92, Happy to hear you're enjoying remult.

While we are still working on a more robust DB versioning/migration solution, you can use the current functionality configured using the `ensureSchema` option in `RemultServerOptions` sent to `remultExpress`.

When the `ensureSchema` option is set to true (the default), as the api starts, it compares the current state of the database with the entities in the code - and adds any missing entity or entity column.
This approach balances ease of use and data integrity - no data will ever be lost and no column/entity will be dropped unintentionally. It is also a great strategy for forward and backward compatability of the db.

For db versioning, until we'll add one built into remult, I use the following code to perform changes as the application evolves:
```ts
import { Entity, Fields, IdEntity, remult, SqlDatabase } from "remult";
import { recordChanges } from "../app/change-log/change-log";
import { Computer } from "../app/computers/computer";

@Entity(undefined!, {
  dbName: "versionInfo"
})
export class VersionInfo extends IdEntity {
  @Fields.number()
  version: number = 0;
}

export async function versionUpdate() {
  let version = async (ver: number, what: () => Promise<void>) => {
    let v = await remult.repo(VersionInfo).findFirst();
    if (!v) {
      v = remult.repo(VersionInfo).create();
      v.version = 0;
    }
    if (v.version <= ver - 1) {
      await what();
      v.version = ver;
      await v.save();
    }
  };

  await version(1, async () => {
    for await (const c of await remult.repo(Computer).query()) {
      await recordChanges(c, {
        forceDate: c.createDate,
        forceNew: true
      })
    }
  });
  await version(2, async () => {
    const db = remult.dataProvider as SqlDatabase;
    const computer = remult.repo(Computer).metadata;
    const indexColumns = await Promise.all(
      [computer.fields.origin, computer.fields.createDate]
        .map(c => c.getDbName()))
    await db.execute(`create index if not exists computerIndex on ${await computer.getDbName()}  (${indexColumns.join(',')})`)
  });
}
``` 

> In the next version of remult the usage of `getDbName` will be replaced with an easier implementation

As for production use, remult is used in production in multiple project.
The oldest project that uses remult for more than 4 years now is the [Hagai Food Distribution app`](https://noam-honig.github.io/salmaz/en.html) that is used by thousands of volunteers to deliver ~400K food parcels a year to families in need.

Here is it's [source code](https://www.github.com/noam-honig/food-basket-delivery)

Remult is used in multiple other projects, some are open source and on [Git Hub](https://github.com/remult/remult/network/dependents?dependents_after=MjQwNzkzNzE5MjY), and others are on private repos built by commercial companies.

**Comment by noam-honig** [comment]
We've just released version https://github.com/remult/remult/releases/tag/v0.26.0 with a built in solution for db migrations:
https://remult.dev/docs/migrations.html

Check it out
